File: src/middleware/rateLimit.ts
import { Request, Response, NextFunction } from "express";
import { getRedisClient } from "../lib/redis";
import { getEnv } from "../config/env";
import { logger } from "../utils/logger";
import { GatewayError } from "../types";

const RATE_LIMIT_PREFIX = "rate_limit:";

export async function rateLimitMiddleware(
  req: Request,
  res: Response,
  next: NextFunction,
): Promise<void> {
  const env = getEnv();
  const maxRequests = env.RATE_LIMIT_PER_IP;
  const windowMs = env.RATE_LIMIT_WINDOW_MS;

  // Get client IP
  const ip = req.ip || req.socket.remoteAddress || "unknown";
  const key = `${RATE_LIMIT_PREFIX}${ip}`;

  const redis = getRedisClient();
  const now = Date.now();
  const windowStart = now - windowMs;

  try {
    // Remove old entries outside the window
    await redis.zremrangebyscore(key, 0, windowStart);

    // Count current requests in window
    const count = await redis.zcard(key);

    if (count >= maxRequests) {
      const error: GatewayError = {
        type: "rate_limit_error",
        code: "RATE_LIMIT_EXCEEDED",
        message: `Rate limit exceeded. Maximum ${maxRequests} requests per ${windowMs / 1000} seconds.`,
        request_id: req.requestId,
      };
      res.status(429).json({ error });
      return;
    }

    // Add current request
    await redis.zadd(key, now, `${now}-${Math.random()}`);

    // Set expiry on the key
    await redis.expire(key, Math.ceil(windowMs / 1000));

    // Add rate limit headers
    res.setHeader("X-RateLimit-Limit", maxRequests.toString());
    res.setHeader(
      "X-RateLimit-Remaining",
      (maxRequests - count - 1).toString(),
    );

    next();
  } catch (err) {
    // If Redis fails, allow the request but log it
    logger.error({
      event: "rate_limit_check_failed",
      error: err instanceof Error ? err.message : String(err),
    });
    next();
  }
}

File: internal/middleware/rate_limit.go
package middleware

import (
	"net/http"
	"strconv"
	"time"

	"github.com/abdo-355/llm-gateway/internal/config"
	"github.com/abdo-355/llm-gateway/internal/db"
	"github.com/abdo-355/llm-gateway/internal/logger"
	"github.com/abdo-355/llm-gateway/internal/metrics"
	"github.com/gin-gonic/gin"
	"github.com/redis/go-redis/v9"
)

type RateLimiter struct {
	redis       *redis.Client
	maxRequests int
	windowMs    int
}

func NewRateLimiter(redis *redis.Client) *RateLimiter {
	env := config.GetEnv()
	return &RateLimiter{
		redis:       redis,
		maxRequests: env.RateLimitPerIP,
		windowMs:    env.RateLimitWindowMs,
	}
}

func (rl *RateLimiter) RateLimit() gin.HandlerFunc {
	return func(c *gin.Context) {
		if c.Request.URL.Path == "/health" {
			c.Next()
			return
		}

		clientIP := c.ClientIP()
		ctx := c.Request.Context()

		key := db.GetRedisKey("ratelimit:" + clientIP)
		window := time.Duration(rl.windowMs) * time.Millisecond

		now := float64(time.Now().UnixMilli())
		cutoff := float64(time.Now().Add(-window).UnixMilli())

		// Step 1: Clean old entries and check count (without adding)
		checkPipe := rl.redis.Pipeline()
		checkPipe.ZRemRangeByScore(ctx, key, "0", strconv.FormatFloat(cutoff, 'f', 0, 64))
		checkPipe.ZCard(ctx, key)

		results, err := checkPipe.Exec(ctx)
		if err != nil {
			logger.Error().
				Str("type", "middleware").
				Str("event", "ratelimit.check_failed").
				Err(err).
				Msg("Rate limit check failed")
			c.Next()
			return
		}

		currentCount := results[1].(*redis.IntCmd).Val()

		if currentCount >= int64(rl.maxRequests) {
			metrics.RateLimitRejectionsTotal.Inc()
			c.AbortWithStatusJSON(http.StatusTooManyRequests, gin.H{
				"error": gin.H{
					"type":    "rate_limit_error",
					"code":    "RATE_LIMIT_EXCEEDED",
					"message": "Rate limit exceeded. Please try again later.",
				},
			})
			return
		}

		// Step 2: Only add the request if allowed
		addPipe := rl.redis.Pipeline()
		addPipe.ZAdd(ctx, key, redis.Z{Score: now, Member: now})
		addPipe.Expire(ctx, key, window)
		if _, err := addPipe.Exec(ctx); err != nil {
			logger.Error().
				Str("type", "middleware").
				Str("event", "ratelimit.record_failed").
				Err(err).
				Msg("Rate limit record failed")
		}

		c.Next()
	}
}
